### ISSUE TITLE
AnyStr becomes unknown

### ISSUE THREAD

**Issue by Avasam** [microsoft/pylance-release#2078]
## Environment data

-   Language Server version: v2021.11.1
-   OS and version: 10.0.19043
-   Python version (& distribution if applicable, e.g. Anaconda): 3.9

## Expected behaviour

Type of value, after the following ternary condition, to be `str` (or some subset/union)

## Actual behaviour

It is `Unknown | Literal`

## Logs

```
Experiment 'pythonaacf' is active
Experiment 'pythonTensorboardExperiment' is active
Experiment 'PythonPyTorchProfiler' is active
Experiment 'pythonDeprecatePythonPath' is active
Experiment 'pythonRunFailedTestsButtonDisplayed' is active
Experiment 'pythonRefreshTestsButtonDisplayed' is active
Experiment 'pythonRememberDebugConfig' is active
> conda info --json
> "C:\Program Files\Python39\python.exe" ~\.vscode\extensions\ms-python.python-2021.11.1422169775\pythonFiles\interpreterInfo.py
Python interpreter path: C:\Program Files\Python39\python.exe
Starting Pylance language server.
> conda --version
> "C:\Program Files\Python39\python.exe" -m flake8 --format=%(row)d,%(col)d,%(code).1s,%(code)s:%(text)s ~\Documents\Git\Auto-Split\src\AutoSplit.py
cwd: ~\Documents\Git\Auto-Split
> "C:\Program Files\Python39\python.exe" -m pylint --msg-template='{line},{column},{category},{symbol}:{msg} --reports=n --output-format=text ~\Documents\Git\Auto-Split\src\AutoSplit.py
cwd: ~\Documents\Git\Auto-Split
> "C:\Program Files\Python39\python.exe" -m mypy --follow-imports=silent --ignore-missing-imports --show-column-numbers --no-pretty ~\Documents\Git\Auto-Split\src\AutoSplit.py
cwd: ~\Documents\Git\Auto-Split
##########Linting Output - flake8##########
##########Linting Output - pylint##########
##########Linting Output - mypy##########
```

## Code Snippet / Additional information

```python
from typing import Optional
import os


last_successfully_loaded_settings_file_path: Optional[str] = None

settings_file_name = "Untitled" \
    if last_successfully_loaded_settings_file_path is None \
    else os.path.basename(last_successfully_loaded_settings_file_path)
```
![image](https://user-images.githubusercontent.com/1350584/142027997-981a30b6-f96c-4d15-87ec-2429b45fdf12.png)
![image](https://user-images.githubusercontent.com/1350584/142028634-e155c6d8-fd8c-4a55-92e0-1d327941c1f1.png)

**Comment by erictraut** [comment]
I cannot repro the problem with the provided code snippet. Please provide a self-contained minimal sample that exhibits the problem you're seeing.

**Comment by Avasam** [comment]
> I cannot repro the problem with the provided code snippet. Please provide a self-contained minimal sample that exhibits the problem you're seeing.

Hi erictraut. I have updated the code snippet to be a totally self-contained example.

Edit: After further review, I realize what was wrong: In my example I needed to explicitely type `last_successfully_loaded_settings_file_path` as `Optional[str]` because it thought it's always `None` (meaning `Never` was passed to `os.path.basename`, resulting in the `Unknown I was seeing`. The updated code snippet shows that. But if you randomize the value to be either `None` or a `str` it'll work correctly.  
So this is probably working as intended, my bad. (feel free to close)
